File: src/backend/routes-api/app/utils/get_distance_between_geo_points.py
import math
from app.consts import EARTH_RADIUS_KM

def get_distance_between_geo_points(lat1: float, long1: float, lat2: float, long2: float) -> float:
    """
    Calcula la distancia en metros entre dos puntos geográficos utilizando la fórmula del haversine.

    :param lat1: Latitud del primer punto.
    :param long1: Longitud del primer punto.
    :param lat2: Latitud del segundo punto.
    :param long2: Longitud del segundo punto.
    :return: Distancia en metros entre los dos puntos.
    """
    R = EARTH_RADIUS_KM * 1000  # Radio de la Tierra en metros

    d_lat = math.radians(lat2 - lat1)
    d_lon = math.radians(long2 - long1)

    a = (math.sin(d_lat / 2) ** 2 +
         math.cos(math.radians(lat1)) *
         math.cos(math.radians(lat2)) *
         math.sin(d_lon / 2) ** 2)

    c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))
    distance = R * c

    return distance

File: src/backend/main-api/utils/getDistanceBetweenGeoPoints.ts
/**
 * Calcula la distancia en metros entre dos puntos geográficos utilizando la fórmula del haversine.
 * @param lat1 Latitud del primer punto.
 * @param long1 Longitud del primer punto.
 * @param lat2 Latitud del segundo punto.
 * @param long2 Longitud del segundo punto.
 * @return Distancia en metros entre los dos puntos.
 */
const getDistanceBetweenGeoPoints = ({
    lat1,
    long1,
    lat2,
    long2,
}: {
    lat1: number;
    long1: number;
    lat2: number;
    long2: number;
}): number => {
    const R = 6371000; // Radio de la Tierra en metros

    const dLat = ((lat2 - lat1) * Math.PI) / 180;
    const dLon = ((long2 - long1) * Math.PI) / 180;

    const a =
        Math.sin(dLat / 2) * Math.sin(dLat / 2) +
        Math.cos((lat1 * Math.PI) / 180) *
            Math.cos((lat2 * Math.PI) / 180) *
            Math.sin(dLon / 2) *
            Math.sin(dLon / 2);

    const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
    const distance = R * c;

    return distance;
};

export default getDistanceBetweenGeoPoints;